### PR TITLE
fix e2e hostdir test

### PR DESCRIPTION
**Pull request by dalanlan**
Test host volume r/w with different containers within a pod
CAUTION: didn't run local test for our cluster is tore down unexpectedly. (The main idea passed the test before on `ubuntu` though.)
Signed-off-by: He Simei hesimei@zju.edu.cn

### CODE REVIEW

**Comment by dalanlan** [comment]
@pmorie 
Would you mind kindly reviewing my code plz:)

**Comment by pmorie** [comment]
@dalanlan Apologies for not reviewing this sooner; you will need to rebase against master.

File: contrib/for-tests/mount-tester/mt.go [hunk 1]
```diff
@@ -152,3 +160,41 @@ func readWriteNewFile(path string) error {
...
+	var content []byte
+	//Expected content "mount-tester new file\n", length (22), retry during 240 second
+	content = TestFileContent(path, 22, 240)
+
+	fmt.Printf("content of file %q: %v\n", path, string(content))
+
+	return nil
+}
+
+func TestFileContent(filePath string, bytesExpected int, retryDuration int) []byte {
```
Comment (pmorie) [inline contrib/for-tests/mount-tester/mt.go]: @dalanlan This function doesn't need to be exported, rename to `testFileContent`

File: contrib/for-tests/mount-tester/mt.go [hunk 2]
```diff
@@ -152,3 +160,41 @@ func readWriteNewFile(path string) error {
...
+				fmt.Printf("Unexpected length of file: found %d, expected %d.Retry", len(contentBytes), bytesExpected)
+			}
+		} else {
+			fmt.Printf("Error read file %s: %v, retry", filePath, err)
+		}
+
+	}
+
+	return contentBytes
+
```
Comment (pmorie) [inline contrib/for-tests/mount-tester/mt.go]: spurious newline, plz remove

File: contrib/for-tests/mount-tester/mt.go [hunk 3]
```diff
@@ -152,3 +160,41 @@ func readWriteNewFile(path string) error {
...
 	return readFileContent(path)
 }
+
+func readFileContentInLoop(path string) error {
+	if path == "" {
+		return nil
+	}
+	var content []byte
+	//Expected content "mount-tester new file\n", length (22), retry during 240 second
+	content = TestFileContent(path, 22, 240)
```
Comment (pmorie) [inline contrib/for-tests/mount-tester/mt.go]: @dalanlan How did you decide on 4 minutes for the retryDuration?

**Comment by pmorie** [comment]
@dalanlan You should bump the version in the mount-test image Makefile to be 0.3

File: contrib/for-tests/mount-tester/mt.go (hunk 3 above)
Comment (dalanlan) [inline contrib/for-tests/mount-tester/mt.go]: Just a simple estimation. What do u have in mind? :)

**Comment by dalanlan** [comment]
Updated. @pmorie

File: contrib/for-tests/mount-tester/mt.go (hunk 3 above)
Comment (pmorie) [inline contrib/for-tests/mount-tester/mt.go]: I'm not sure -- just seems like a really long window to wait if something goes bad.

File: test/e2e/util.go [hunk 4]
```diff
@@ -728,14 +728,14 @@ func runKubectl(args ...string) string {
 }
 
 // testContainerOutput runs testContainerOutputInNamespace with the default namespace.
-func testContainerOutput(scenarioName string, c *client.Client, pod *api.Pod, expectedOutput []string) {
-	testContainerOutputInNamespace(scenarioName, c, pod, expectedOutput, api.NamespaceDefault)
+func testContainerOutput(scenarioName string, c *client.Client, pod *api.Pod, containerId int, expectedOutput []string) {
```
Comment (pmorie) [inline test/e2e/util.go]: I think this test API change should be split into a separate PR -- it has merit on its own.

**Comment by pmorie** [comment]
Close to LGTM but I would prefer the testContainerOutputInNamespace change to be a separate PR.

**Comment by dalanlan** [comment]
I updated the code and could use some help here.
Is Shippable unhappy now? @pmorie

**Comment by pmorie** [comment]
@dalanlan Will take a look.

File: contrib/for-tests/mount-tester/mt.go (hunk 3 above)
Comment (pmorie) [inline contrib/for-tests/mount-tester/mt.go]: @dalanlan On thinking about it, why don't you make the duration of the retry a cli parameter?

File: test/e2e/host_path.go [hunk 5]
```diff
@@ -68,10 +48,10 @@ var _ = Describe("hostDir", func() {
 			fmt.Sprintf("--fs_type=%v", volumePath),
 			fmt.Sprintf("--file_mode=%v", volumePath),
 		}
-		testContainerOutputInNamespace("emptydir r/w on tmpfs", c, pod, 0, []string{
+
+		f.TestContainerOutput("hostdir r/w on tmpfs", pod, 0, []string{
```
Comment (pmorie) [inline test/e2e/host_path.go]: This test case doesn't really can anything to do with tmpfs.  I suggest calling it 'file mode on hostdir' or something similar.

File: test/e2e/host_path.go [hunk 6]
```diff
@@ -87,17 +67,25 @@ var _ = Describe("hostDir", func() {
 			fmt.Sprintf("--rw_new_file=%v", filePath),
 			fmt.Sprintf("--file_mode=%v", filePath),
 		}
-		testContainerOutputInNamespace("emptydir r/w on tmpfs", c, pod, 0, []string{
+		f.TestContainerOutput("hostdir r/w on tmpfs", pod, 0, []string{
```
Comment (pmorie) [inline test/e2e/host_path.go]: Same thing as above; this doesn't really have anything to do with tmpfs.  Rename the `It` argument to something like 'should support r/w' and take the reference to tmpfs out of the first argument to `TestContainerInOutput`.

**Comment by pmorie** [comment]
@dalanlan Basically one nit in the test-case names and arguments, LGTM otherwise.  Fix this and I will tag for merge.

**Comment by pmorie** [comment]
@dalanlan The shippable build failed.  Relevant error message:

```
 GOPATH=$PWD/Godeps/_workspace:$GOPATH go install ./...
# github.com/GoogleCloudPlatform/kubernetes/contrib/for-tests/mount-tester
contrib/for-tests/mount-tester/mt.go:170: undefined: TestFileContent
```

File: contrib/for-tests/mount-tester/mt.go (hunk 3 above)
Comment (dalanlan) [inline contrib/for-tests/mount-tester/mt.go]: I don't mind but i coud not see why we trouble doing that. 
Our main concern is to set up a proper waiting time, changing it as a parameter won't fix that after all. Giving a comfortable waiting time should do.

**Comment by dalanlan** [comment]
@pmorie 
Thx so much for your kind help.
I updated the code and it failed again. Had no idea why i cannot access any detailed information of Shippable all around (It always showed me 'Build not found').
And I didn't make the duration of the retry a cli parameter yet. Let me know if you insist:)

**Comment by pmorie** [comment]
@dalanlan Sorry for the latency, I have been out of the office this week.  Looking at your shippable issue now.

**Comment by pmorie** [comment]
@dalanlan Hrm, I thought I commented with the failure in shippable:

```
 GOPATH=$PWD/Godeps/_workspace:$GOPATH go install ./...
# github.com/GoogleCloudPlatform/kubernetes/contrib/for-tests/mount-tester
contrib/for-tests/mount-tester/mt.go:170: undefined: TestFileContent
```

**Comment by dalanlan** [comment]
@pmorie  
Uhh...It's weird. 
I did update the code as you could see.

**Comment by dalanlan** [comment]
I commited the code again. Should be ok to test now.  @pmorie 

(The check still seems super weird, though.)

**Comment by pmorie** [comment]
@dalanlan OK, I will pull this down and run the test tomorrow myself.  I need to build the latest mounttest for this, right?

**Comment by pmorie** [comment]
@k8s-bot retest please

**Comment by dalanlan** [comment]
yep:)

**Comment by dalanlan** [comment]
"retest" seems not working.
"ok to test" may work?
@pmorie

**Comment by pmorie** [comment]
@dalanlan I don't expect this to work until the new versions of the images you made are pushed

File: contrib/for-tests/mount-tester/mt.go [hunk 7]
```diff
@@ -152,3 +160,41 @@ func readWriteNewFile(path string) error {
...
 }
+
+func readFileContentInLoop(path string) error {
+	if path == "" {
+		return nil
+	}
+	var content []byte
+	//Expected content "mount-tester new file\n", length (22), 
+	//retry during 240 second
+	content = testFileContent(path, 22, 240)
```
Comment (pmorie) [inline contrib/for-tests/mount-tester/mt.go]: @dalanlan I think 240 needs to be a CLI param to the mount tester, and '22' ought to be a constant

File: test/e2e/host_path.go [hunk 8]
```diff
@@ -20,42 +20,22 @@ import (
...
 	"os"
 	"path"
 
 	. "github.com/onsi/ginkgo"
-	. "github.com/onsi/gomega"
 )
 
 //TODO : Consolodate this code with the code for emptyDir.
 //This will require some smart.
 var _ = Describe("hostDir", func() {
```
Comment (pmorie) [inline test/e2e/host_path.go]: @dalanlan s/hostDir/hostPath/

**Comment by dalanlan** [comment]
Thx. Updated :)
@pmorie

**Comment by erictune** [comment]
@thockin confirm that this test is not critical for v1?

**Comment by thockin** [comment]
not critical for 1.0

**Comment by pmorie** [comment]
@dalanlan Would you rebase this and we can get it in?

**Comment by dalanlan** [comment]
@pmorie  Rebased and squashed.

**Comment by pmorie** [comment]
@dalanlan We should updated `mounttest-user` too.

**Comment by dalanlan** [comment]
PTAL

**Comment by pmorie** [comment]
@k8s-bot ok to test (though I do not expect it to work)

**Comment by pmorie** [comment]
@dalanlan I'm testing it out locally now

**Comment by pmorie** [comment]
@dalanlan I was testing this locally and I think there are a couple of things that need to be fixed.  Will comment.

File: test/e2e/host_path.go [hunk 9]
```diff
@@ -77,27 +76,40 @@ var _ = Describe("hostPath", func() {
...
 		}, namespace.Name,
 		)
+
+		pod.Spec.Containers[1].Args = []string{
+			fmt.Sprintf("--file_content_in_loop=%v", filePath),
+			fmt.Sprintf("--retry_time=%d", retryDuration),
+		}
+		//Read the content of the file with the second container to
+		//verify volumes  being shared properly among continers within the pod.
+		testContainerOutputInNamespace("hostPath r/w", c, pod, 1, []string{
```
Comment (pmorie) [inline test/e2e/host_path.go]: Calling `testContainerOutputInNamespace` twice will run this pod twice -- is that what you intended to do?

File: test/e2e/host_path.go [hunk 10]
```diff
@@ -77,27 +76,40 @@ var _ = Describe("hostPath", func() {
...
 
 		pod.Spec.Containers[0].Args = []string{
-			fmt.Sprintf("--fs_type=%v", volumePath),
 			fmt.Sprintf("--rw_new_file=%v", filePath),
 			fmt.Sprintf("--file_mode=%v", filePath),
 		}
+
 		testContainerOutputInNamespace("hostPath r/w", c, pod, 0, []string{
 			"mode of file \"/test-volume/test-file\": -rw-r--r--",
 			"content of file \"/test-volume/test-file\": mount-tester new file",
```
Comment (pmorie) [inline test/e2e/host_path.go]: I would expect this not to work since the container isn't run with `--file-content`.

File: test/e2e/host_path.go (hunk 9 above)
Comment (pmorie) [inline test/e2e/host_path.go]: If you want to test both the cases of whether a container can read the file it just wrote, and whether another container in the same pod can see the file, you should break them into two separate cases.

File: test/e2e/host_path.go (hunk 10 above)
Comment (pmorie) [inline test/e2e/host_path.go]: Disregard, I had the code incorrect in my head :)

**Comment by pmorie** [comment]
@dalanlan I ran into some failures testing this locally -- digging into them now.

File: contrib/for-tests/mount-tester/mt.go [hunk 11]
```diff
@@ -42,6 +46,9 @@ func init() {
 	flag.StringVar(&newFilePath0644, "new_file_0644", "", "Path to write to and read from with perm 0644")
 	flag.StringVar(&newFilePath0666, "new_file_0666", "", "Path to write to and read from with perm 0666")
 	flag.StringVar(&newFilePath0777, "new_file_0777", "", "Path to write to and read from with perm 0777")
+	flag.StringVar(&readWriteNewFilePath, "rw_new_file", "", "Path to write to and read from")
+	flag.StringVar(&readFileInLoopContentPath, "file_content_in_loop", "", "Path to read the file content in loop from")
```
Comment (pmorie) [inline contrib/for-tests/mount-tester/mt.go]: @dalanlan The `readWriteNewFilePath` variable is never used to drive creating the new file.  Perhaps you lost something in your rebase?  I'm not sure that we need this now that there are other options for creating new files with different permissions.

**Comment by dalanlan** [comment]
Sorry, updated the code.
@pmorie  plz take a double check.

**Comment by dalanlan** [comment]
@pmorie  Got tired of rebase work. Didnt mean to rush you, yet PTAL once you're available :)

**Comment by pmorie** [comment]
@dalanlan I'll pull it and test it out now.

**Comment by pmorie** [comment]
Okay, I got a successful run of this locally.  Of course, I had to make SELinux permissive, but that is expected.

**Comment by pmorie** [comment]
@dalanlan One thing that should be changed (this is my fault, not yours): in `test/images/mount-tester-user`, the `Dockerfile` should reference `google_containers`, not `google-containers`.  If you fix that, it should be good to go.

**Comment by pmorie** [comment]
Otherwise this LGTM.  I'll re-test and label this PR once that last nit is fixed.  Thanks for adding this test :)

**Comment by pmorie** [comment]
Actually I will just fix that, and you can go in on top.

**Comment by pmorie** [comment]
Opened #12935

**Comment by dalanlan** [comment]
Updated. @pmorie

**Comment by pmorie** [comment]
I got a clean run of these tests locally.  I don't expect the jenkins tests to pass until the new images are built.

Adding LGTM.  Whoever merges this will need to ensure the images are built and pushed to gcr.io.

**Comment by zmerlynn** [comment]
@k8s-bot test this

**Comment by zmerlynn** [comment]
(`mounttest:0.4` and `mounttest-user:0.2` pushed, so this should pass now)
